src/scores/continuous/qq_impl.py

```diff
@@ -0,0 +1,134 @@
+"""
+
+from typing import Dict, Iterable
+
+import numpy as np
+import xarray as xr
+
+from scores.processing import broadcast_and_match_nan
+from scores.typing import FlexibleDimensionTypes, XarrayLike, all_same_xarraylike
+from scores.utils import gather_dimensions
+
+NP_INTERP_METHODS = [
+    "inverted_cdf",
+    "averaged_inverted_cdf",
+    "closest_observation",
+    "interpolated_inverted_cdf",
+    "hazen",
+    "weibull",
+    "linear",
+    "median_unbiased",
+    "normal_unbiased",
+    "lower",
+    "higher",
+    "midpoint",
+    "nearest",
+]
+
+
+def generate_qq_plot_data(
+    fcst: XarrayLike,
+    obs: XarrayLike,
+    *,
+    quantiles: Iterable[float],
+    interpolation_method: str = "linear",
+    reduce_dims: FlexibleDimensionTypes | None = None,
+    preserve_dims: FlexibleDimensionTypes | None = None,
+) -> XarrayLike:
+    """
+    Calculate the quantiles of empirical forecast and observation data for a
+    quantile-quantile (Q-Q) plot.
+
+    Forecast and observation data are broadcast and NaN values are matched
+    before calculating the quantiles. This function calculates the quantiles
+    for the user to plot using their preferred plotting library.
+
+    Currently this function is implemented for working with raw forecast and
+    observation data rather than a theoretical probability distribution.
+
+    Args:
+        fcst: Forecast data.
+        obs: Observation data.
+        quantiles: Quantiles to compute, e.g., [0.1, 0.5, 0.9].
+        interpolation_method: Interpolation method for estimating the quantile value,
+            default is 'linear'.
+            Valid options include 'inverted_cdf', 'averaged_inverted_cdf',
+            'closest_observation', 'interpolated_inverted_cdf', 'hazen',
+            'weibull', 'linear', 'median_unbiased', 'normal_unbiased',
+            'lower', 'higher', 'midpoint', and 'nearest'. See :py:func:`numpy.quantile`
+            for more information on interpolation method options.
+        reduce_dims: Dimensions to reduce over when calculating quantiles.
+        preserve_dims: Dimensions to preserve when calculating quantiles.
+
+    Returns:
+        An xarray object with a new dimension called "data_source".
+
+    Raises:
+        ValueError: If the interpolation method is invalid.
+        ValueError: if quantiles are not between 0 and 1.
+        ValueError: If dimensions named 'data_source' are present in the input data.
+        ValueError: If fcst and obs are Datasets with different data variables.
+        ValueError: If a user tries to preserve all dimensions.
+        TypeError: If fcst and obs are not both xarray DataArrays or Datasets.
+
+    References:
+        Déqué, M. (2012). Deterministic forecasts of continuous variables. In I. T.
+        Jolliffe & D. B. Stephenson (Eds.), Forecast verification: A practitioner’s
+        guide in atmospheric science (2nd ed., pp. 77–94). Wiley. https://doi.org/10.1002/9781119960003
```

Comment:
```suggestion
        Déqué, M. (2012). Deterministic forecasts of continuous variables. In I. T.
        Jolliffe & D. B. Stephenson (Eds.), Forecast verification: A practitioner’s
        guide in atmospheric science (2nd ed., pp. 77–94). Wiley. https://doi.org/10.1002/9781119960003.ch5
```
As above, I think this is the DOI for the specific chapter (but please confirm)